ARM opcode & operand encoding information (based on auto-sync)

AngelDev06 opened the pull request:
I implemented two new structures.
The first one being:
![Screenshot 2023-06-16 051447](https://github.com/capstone-engine/capstone/assets/134562527/728ad2ce-5e87-40b7-b02d-281f85ebfc46)
and the second one being:
![Screenshot 2023-06-16 051641](https://github.com/capstone-engine/capstone/assets/134562527/85411cc8-4e08-44f5-8575-12a2f9fa1ed9)
where both are defined in capstone.h and are available for use by all auto-sync archs. An instance of the first struct is provided as part of the detail member as shown here: 
![Screenshot 2023-06-16 052108](https://github.com/capstone-engine/capstone/assets/134562527/a98230a1-5b87-4098-9b6d-1c952073b2b2)
while the instance of the second struct is provided directly as part of the arch's operand (detail only) such as `cs_arm_op`. Currently only the **ARM** arch fills these structs with info as this is the only arch I have generated the required tables for. However this is entirely based on auto-sync so once I publish my modified copy of llvm-capstone (containing the function used to generate the encoding strings), it should be easy to generate the tables for the rest of the auto-sync archs. 

- The info for the opcode encoding is mapped from the `ARMGenCSMappingInsn.inc` file (and is also wrapped under CAPSTONE_DIET) in the function `ARM_set_instr_map_data` (which is invoked right after the instruction is decoded). 
- The info for the operand however, is mapped from the `ARMGenCSMappingInsnOp.inc` file (like the rest of the operand info) and only after `ARM_add_cs_detail` is called. The mapping operation takes place on functions such as `ARM_set_detail_op_reg` and `ARM_set_detail_op_imm` and in some special cases it's hardcoded inside the rest of the `add_cs_detail` functions that are implemented (such as `add_cs_detail_general`). Note that for the reglist operand, since capstone breaks it down to a list of register operands (each one being a seperate `cs_arm_op`), I proceeded to do the same for the encoding. I added a useful comment in that part.

A test I did myself on a few instructions to verify that it works includes the following output:
![Screenshot 2023-06-16 044229](https://github.com/capstone-engine/capstone/assets/134562527/eec6050f-1ba0-4c94-b501-93a443163c03)
![Screenshot 2023-06-16 042025](https://github.com/capstone-engine/capstone/assets/134562527/2a357c46-2a2b-4370-bcad-ba84055c9368)
and yes I verified the results from the official **ARM** documentation. More info:

- [LDM](https://developer.arm.com/documentation/ddi0406/c/Application-Level-Architecture/Instruction-Details/Alphabetical-list-of-instructions/LDM-LDMIA-LDMFD--ARM-) ([encoding](https://documentation-service.arm.com/static/5f8db1f7f86e16515cdba472?token=))
- [VADD.F32](https://developer.arm.com/documentation/ddi0406/c/Application-Level-Architecture/Instruction-Details/Alphabetical-list-of-instructions/VADD--floating-point-) ([encoding](https://documentation-service.arm.com/static/5f8db1f7f86e16515cdba44d?token=))
- [STR](https://developer.arm.com/documentation/ddi0406/c/Application-Level-Architecture/Instruction-Details/Alphabetical-list-of-instructions/STR--immediate--ARM-) ([encoding](https://documentation-service.arm.com/static/5f8db1f7f86e16515cdba394?token=))

Let me know if there are any issues.

b1llow committed:
ST SUB SYSCALL XOR XNOR

b1llow committed:
update `TriCore*.inc`

b1llow committed:
add FPU Instructions

b1llow committed:
fix build

b1llow committed:
`Decode*Instruction`

b1llow committed:
fix `DecodeInstruction`

b1llow committed:
add `tricore` to `cstool`

b1llow committed:
add `tricore` to `cstest`

b1llow committed:
fix

b1llow committed:
fix

b1llow committed:
- add `tricore` to python binding
- try fix `test_corpus.py`

b1llow committed:
add more `tricore` cstest

b1llow committed:
add `sign_ext` `zero_ext`

b1llow committed:
fix

b1llow committed:
fix `DecodeRegisterClass`
- `DecodeABSInstruction`
- `DecodeRLCInstruction`
- `DecodeSSRInstruction`
- `DecodeSRCInstruction`

b1llow committed:
fix `DecodeRegisterClass`

b1llow committed:
fix `BO` `BOL`

b1llow committed:
fix `SRRS`

b1llow committed:
fix

b1llow committed:
fix `RRPW`

b1llow committed:
fix `INSERT_rrpw`

b1llow committed:
fix `RR`

b1llow committed:
fix

b1llow committed:
fix `mov.a` `mov.d`

b1llow committed:
fix `imask`

b1llow committed:
Whether to call `a10` a `sp`

b1llow committed:
add `MTCR`

b1llow committed:
add `BO`

b1llow committed:
docs: Refactor project structure and remove unnecessary file.

b1llow committed:
fix: TriCore architecture disassembly codes

- Rename `ISLR_post_increment` to `ISLR_pos` for clarity
- Fix register decoding for TriCore architecture in `TriCoreDisassembler.c`
- Add new file `LoadStore.s.cs` to `suite/MC/TriCore`

b1llow committed:
fix: TriCore instruction decoding and printing.

- Modify TriCore instructions to use bracket syntax and offsets for better clarity
- Add off4_fixup and printSExtImm_n

b1llow committed:
fix: decode error

b1llow committed:
Fix: TriCore instruction operations and decoding.

b1llow committed:
fix: TriCore Instruction Formats and Printing

- Add new operand type `off18imm`
- Add `printOff18Imm` function for printing specific immediate value

b1llow committed:
refactor: Refactor TriCore instructions and operands

- Add new immediate operands and refactor code for better readability in TriCoreInstrInfo
- Update InstPrinter functions to handle new immediate operands and remove unused function

b1llow committed:
feat: Fix bugs and update instructions for TriCore architecture.

- Fix bug in `printDisp8Imm` function in `TriCoreInstPrinter.c`
- Add new `CALL_sb` instruction to `TriCoreInstrInfo.td`
- Reorder instruction definitions and operands in `TriCoreInstrInfo.td`

b1llow committed:
fix: decode `j` `call` `loop`

b1llow committed:
unique tests

b1llow committed:
fix: `disp` print and fill

b1llow committed:
just add TriCoreISA enum

b1llow committed:
add tricore Predicates

b1llow committed:
cleanup

b1llow committed:
add some tricore v1.1 inst

b1llow committed:
fix `cachea`

b1llow committed:
refactor: Refactor TriCore register names.

- Rename TriCore register names for better readability
- Update TriCore instruction information

b1llow committed:
refactor: Improve TriCore instruction handling.

- Add new instructions for TriCore architecture
- Implement changes to TriCoreInstrInfo.td

b1llow committed:
refactor: Add support for new TriCore instructions and constraints.

- Add support for V110 and V120 in various LD instructions
- Define new multiclass for code reuse
- Restructure LD_A_v120 in favor of LD_A with Requires constraint
- Add new defs for LT_U with suffixes for V110 support

b1llow committed:
refactor: Update TriCore instruction requirements

- Update `Requires` directives for several instructions to reflect changes in hardware versions

b1llow committed:
refactor: Refactor TriCore instruction decoding and register definition.

- Update TriCore processor register definitions with auto-generated file `TriCoreGenCSRegEnum.inc`
- Add several new TriCore processor instructions with auto-generated file `TriCoreGenCSInsnEnum.inc`
- Update TriCore_OP_GROUP enumeration with auto-generated file `TriCoreGenCSOpGroup.inc`
- Rename and restructure TriCore processor register classes
- Remove unused register class definitions and related code

b1llow committed:
add `tricore_feature` support

b1llow committed:
feat: Add and remove TriCore instructions.

- Add 3 new TriCore instructions
- Remove TriCore instruction "TriCore_INS_INIT"
- Alphabetized and rearranged various TriCore instructions
- Commented out code remains in the diff but is not part of the program.

b1llow committed:
refactor: Add new TriCore instructions and remove deprecated ones.

- Remove deprecated instructions
- Update comments and formatting

b1llow committed:
refactor: Refactor TriCore instruction definitions and mappings

- Add new `multiclass mI_MADDRsh_` to handle `MADDR_H` and `MADDRS_H` in `arch/TriCore/TriCoreInstrInfo.td`

b1llow committed:
refactor: Improve Architecture Instruction Information.

- Refactor TriCore instruction information file
- Simplify and optimize certain instructions

b1llow committed:
refactor: Refactor TriCore instruction classes and operands

b1llow committed:
Optimize TriCore instruction information.

- Refactor TriCore instructions
- Simplify operand encoding for better readability

b1llow committed:
Optimize TriCore instruction information.

- Refactor and optimize instruction encoding for TriCore architecture
- Improve performance and reduce code size by simplifying opcodes and encoding logic
- Update instruction definitions in TriCoreInstrInfo.td file
- Omit some large but non-essential changes in diff summary for readability

b1llow committed:
refactor: Improve TriCore instruction definitions in architecture file

- Refactor architecture specific code for TriCore
- Update TriCore instruction information in [arch/TriCore/TriCoreInstrInfo.td]
- Improve performance and readability of relevant functions

b1llow committed:
refactor: Improve instruction handling for TriCore architecture

b1llow committed:
Add support for TriCore V162 and new instructions/operands.

- Add new instruction `MOVZ_A`, remove instruction `NOT`, and add several new multiply and multiply-subtract instructions
- Move `multiclass mISR_1` and `multiclass mISYS_0` to separate file and fix typo in `rfe` instruction in `mISYS_0`
- Add support for new CPU feature `TriCore_FEATURE_HasV162` and update relevant inc files.

b1llow committed:
Refactor: Optimize TriCore instruction information.

- Refactor TriCore instruction info
- Improve code organization
- Optimize instruction handling
- Eliminate unnecessary code duplication

b1llow committed:
refactor: Improve TriCore floating-point operations in instruction set

b1llow committed:
refactor: Optimize TriCore instructions in arch/TriCore/TriCoreInstrInfo.td

- Add missing instructions and update existing ones

b1llow committed:
fix

b1llow committed:
feat: Add support for TriCore feature bits and new architectures

- Add support for new Tricore architectures
- Clean up redundant instructions definitions
- Modify architecture options for the TRICORE mode
- Update disallowed modes for Tricore architecture

b1llow committed:
feat: Refactor and improve triCore platform support

b1llow committed:
feat: Update TriCore processor support and architecture modes

- Add support for TRICORE architecture modes 110, 120, 130, 131, 160, 161, and 162 in test_corpus.py
- Change the TriCore mode from `CS_MODE_TRICORE` to `CS_MODE_TRICORE_162` in `Tricore/*.s.cs`

b1llow committed:
fix

b1llow committed:
add some tc162 tests

b1llow committed:
fix TriCoreDisassembler.c from tests

b1llow committed:
fix tricore tests

b1llow committed:
feat: Update Tricore assembly code and disassembler logic.

- Add new Tricore test `tc110.s.cs`

b1llow committed:
add tc110 tests and fix tricore decode

b1llow committed:
fix `CADD` `CSUB`

b1llow committed:
fix tc110 test and fix decode

b1llow committed:
fix `RCR` printer

b1llow committed:
fix

b1llow committed:
fix

b1llow committed:
fix

b1llow committed:
fix tc1.1 tests

b1llow committed:
fix tc1.6.2 tests

b1llow committed:
fix tc1.6.2 tests

b1llow committed:
add tests

b1llow committed:
Update TriCore instructions in TriCoreInstrInfo.td

b1llow committed:
fix tests

b1llow committed:
fix tests

b1llow committed:
fix tests

b1llow committed:
fix `TRICORE_GENERIC` inst

b1llow committed:
fix tests

b1llow committed:
fix `TriCore_getRegisterName`

b1llow committed:
cleanup

b1llow committed:
fix `TRICORE_OP_MEM`

b1llow committed:
fix `TriCoreGenCSMappingInsnName.inc`

b1llow committed:
Update tricore `.inc`

b1llow committed:
fix `test_corpus.py`

b1llow committed:
fix fuzz

b1llow committed:
fix fuzz

b1llow committed:
update TriCoreGenAsmWriter.inc

b1llow committed:
add TriCore in README

b1llow committed:
fix tricore endian

junghee committed:
Add handling of prefix `NOTRACK`

junghee committed:
Merge pull request #1 from junghee/junghee/notrack

Add handling of prefix `NOTRACK`

b1llow committed:
fix: support for TriCore call group mapping

b1llow committed:
fix all tricore compile warnings

raigorx committed:
fix: compile error missing file cstool msvc/visual studio

kabeor committed:
Merge pull request #2000 from raigorx/patch-1

fix: compile error missing file cstool msvc/visual studio

b1llow committed:
Format all .(c|h) code

kabeor committed:
Merge pull request #2002 from junghee/next

b1llow committed:
Fix format error

b1llow committed:
Fix `print_insn_detail_tricore` and disp fill

b1llow committed:
Fix tricore python binding

- fix HACK.TXT

b1llow committed:
Fix HACK.TXT

b1llow committed:
Fix tricore.h and remove `inc` folder

radare committed:
Fix make builds with SH support

b1llow committed:
Fix tricore.h

kabeor committed:
Merge pull request #2006 from trufae/make-sh

b1llow committed:
Merge branch 'next' into tricore

b1llow committed:
Upper all `inc` and fix

b1llow committed:
Upper tricore_const.py

b1llow committed:
Update suite/test_corpus3.py

Co-authored-by: Rot127 <45763064+Rot127@users.noreply.github.com>

peace-maker committed:
RISCV: add more instruction groups

Add call, ret, int and branch_relative instruction groups to riscv
mappings.

trufae committed:
Add missing architectures in the readme and sort them (#2008)

Co-authored-by: pancake <pancake@nopcode.org>

b1llow committed:
Merge branch 'next' into tricore

kabeor committed:
Merge pull request #1973 from imbillow/tricore

aquynh committed:
Update README.md

remove some broken badges

b1llow committed:
Fix compilation warnings related to TriCore. (#2012)

* fix tricore compiler warn `-Wmissing-prototypes`

* include cstool.h

Rot127 committed:
Fix SH build warnings.

Rot127 committed:
Update indentation

kabeor committed:
Merge pull request #2016 from Rot127/fix_sh_warnings

Fix SH build warnings.

b1llow committed:
Fix tricore disp24 print

Rot127 committed:
Fix warnings about unhandled switch cases by adding defaults.

b1llow committed:
Fix tricore disp print

b1llow committed:
Refactor TriCoreInstPrinter.c

b1llow committed:
Update arch/TriCore/TriCoreInstPrinter.c

Co-authored-by: Rot127 <45763064+Rot127@users.noreply.github.com>

b1llow committed:
Update arch/TriCore/TriCoreInstPrinter.c

Co-authored-by: Rot127 <45763064+Rot127@users.noreply.github.com>

b1llow committed:
Update arch/TriCore/TriCoreInstPrinter.c

Co-authored-by: Rot127 <45763064+Rot127@users.noreply.github.com>

kabeor committed:
Merge pull request #2018 from imbillow/tricore-fix-disp

Rot127 committed:
Initialize rw

b1llow committed:
Fix ctest build

b1llow committed:
Fix disp fill

kabeor committed:
Merge pull request #2017 from Rot127/fix_sh_warnings

kabeor committed:
Merge pull request #2019 from imbillow/tricore-fix-disp

stevielavern committed:
Fix register accesses & operands of multiple PAC instructions

b1llow committed:
Fix disassemble of `xor`

- update generate file using https://github.com/Rot127/llvm-capstone/tree/78180b63f827ee38e23375e7a00825e848aa6956

kabeor committed:
Merge pull request #2023 from imbillow/tricore_fix

Fix disassemble of `xor`

gdbinit committed:
Add missing arm64 to Darwin LIBARCHS and remove dependencies tracking if building universal binary (dependencies not compatible with fat binaries)

gdbinit committed:
Fix if indent

kabeor committed:
Merge pull request #2022 from stevielavern/fix_pac

kabeor committed:
Merge pull request #2024 from gdbinit/fat_fix

Add missing arm64 to Darwin LIBARCHS and remove dependencies tracking…

b1llow committed:
Fix compile warning

kabeor committed:
Merge pull request #2027 from imbillow/tc-fix

Rot127 committed:
Set mnemonic in PPC_printInst().

Because the cs_insn->mnemonic gets set *after* the
post_printer is called and cs_insn->mnemonic is not memset to 0
before, post_printer() of PPC receives a cs_insn->mnemonic with
random data.
This in turn leads randomly to incorrect setting of the branch hints
and the update_cr0 flag. Because those are set in the post_printer()
of PPC.

kabeor committed:
Merge pull request #2028 from Rot127/set_cr0_fix

Rot127 committed:
Replace strncpy with memcpy because we copy from larger to smaller buffer.

kabeor committed:
Merge pull request #2033 from Rot127/set_cr0_fix

b1llow committed:
Fix compile warnings under `CMake -Wall`

Rot127 committed:
Pull auto-sync's changes from https://github.com/capstone-engine/capstone/commit/2ab11ad9bd1c7278e2eefd9998100629c3f227c7

b1llow committed:
Add `.clang-format` and format

b1llow committed:
Add operands access support for TriCore

kabeor committed:
Merge pull request #2034 from imbillow/tc-reg-rw

Add operands access support for TriCore

peace-maker committed:
Add ctest regression tests

Check for 32 and 64 bit mode as well as compressed instructions.

peace-maker committed:
Merge branch 'next' into riscv_insn_groups

Rot127 committed:
Hotfix for -DCAPSTONE_DIET build.

AngelDev06 committed:
Added opcode and operand encoding info for ARM

as well as wsl support for the updater

kabeor committed:
Merge pull request #2038 from Rot127/fix_diet_build

peace-maker committed:
Fix Python Bindings after changes to cs_detail

#2034 changed the `regs_read` array size and added a new `writeback`
element to the cs_detail struct.

Those changes weren't reflected in the Python bindings causing details
to be missing.

peace-maker committed:
Fix TriCore Python bindings

The structs were out of sync and were missing a few elements.

The `update_flags` element wasn't exposed at all.

peace-maker committed:
Fix RISCV Python bindings

The structs had changed and the `need_effective_addr` element wasn't exposed.

kabeor committed:
Merge pull request #2041 from peace-maker/fix_python_bindings

kabeor committed:
Merge pull request #2007 from peace-maker/riscv_insn_groups

RISCV: Add call, int and branch_relative instruction groups

Rot127 commented on suite/auto-sync/CppTranslator/CppTranslator.py:
```diff
@@ -75,8 +76,8 @@ class Translator:
     template_collector: TemplateCollector = None
-    src_paths: [Path]
-    out_paths: [Path]
+    src_paths: list[Path]
```
Thanks for these fixes. I'll cherry pick them to the ARM PR

Rot127 commented on suite/auto-sync/CppTranslator/arch_config.json:
```diff
@@ -28,91 +28,91 @@
   "ARM": {
     "files_to_translate": [
       {
-        "in": "{AUTO_SYNC_ROOT}/vendor/llvm_root/llvm/lib/Target/ARM/Disassembler/ARMDisassembler.cpp",
```
Why is `llvm` removed here?

AngelDev06 replied:
It was removed because I updated the auto-sync root path passed to the translator. I noticed that all the files accessed from the updater script and the translator/differ etc. are inside the llvm folder so `{AUTO_SYNC_ROOT}` is a path to the folder (instead of the root directory of llvm-capstone). This wasn't really necessary but I did it to simplify the paths. I should have left a hint for that. (the symlink created also points to that directory)

Rot127 replied:
Ah ok. Guess it is fine then.

Rot127 commented on suite/auto-sync/CppTranslator/Differ.py:
```diff
@@ -634,7 +634,7 @@ def parse_args() -> argparse.Namespace:
     )
     cfg = Configurator(args.arch, args.config_path, args.grammar, args.lang_so)
-
+    
```
Remove the `spaces

Rot127 commented on suite/auto-sync/CppTranslator/CppTranslator.py:
```diff
@@ -466,13 +467,14 @@ def parse_args() -> argparse.Namespace:
         log.fatal("Python >= v3.11 required.")
         exit(1)
-
+    
```
Remove the spaces

Rot127 commented on suite/auto-sync/Update-Arch.sh:
```diff
@@ -111,62 +112,71 @@ tblgen="$2"
-  cp __ARCH__GenCSSystemRegisterEnum.inc $arch"GenCSSystemRegisterEnum.inc"
-  cp __ARCH__GenSystemRegister.inc $arch"GenSystemRegister.inc"
+  mv __ARCH__GenCSSystemRegisterEnum.inc "../$arch""GenCSSystemRegisterEnum.inc"
+  mv __ARCH__GenSystemRegister.inc "$arch""GenSystemRegister.inc"
```
This whole `__ARCH__` and `##ARCH##` replacement is outdated (it was a horrible idea to begin with). The `AArch64` tblgen version now generates the files with proper arch names. You'll see when you rebase.

Rot127 commented on Mapping.c:
```diff
@@ -158,6 +158,18 @@ void map_groups(MCInst *MI, const insn_map *imap)
+void map_opcode_encoding(MCInst *MI, const insn_map *imap)
+{
+#ifndef CAPSTONE_DIET
+	if (!MI->flat_insn->detail)
```
After rebase replace with `if (!detail_is_set(MI))`

Rot127 commented on include/capstone/capstone.h:
```diff
@@ -289,6 +289,36 @@ typedef struct cs_opt_skipdata {
+	uint8_t indexes[64];
+	uint8_t
+		bit_count; ///< Specifies the number of bits that form the full opcode.
+} cs_opcode_encoding;
```
After the rebase please move it to `Mapping.h`

AngelDev06 replied:
Isn't the intention for the struct to be publicly available? I might be wrong but `Mapping.h` isn't included by the user.

Rot127 replied:
You are right. Leave it there.

Rot127 commented on arch/ARM/ARMMapping.c:
```diff
@@ -391,6 +392,7 @@ static void add_cs_detail_general(MCInst *MI, arm_op_group op_group,
 			(ARMCC_CondCodes)MCOperand_getImm(MCInst_getOperand(MI, OpNum));
 		if ((unsigned)CC == 15 && op_group == ARM_OP_GROUP_PredicateOperand) {
 			ARM_get_detail(MI)->cc = ARMCC_UNDEF;
+
```
Remove newline

Rot127 commented on arch/ARM/ARMMapping.c:
```diff
@@ -430,7 +432,7 @@ static void add_cs_detail_general(MCInst *MI, arm_op_group op_group,
 		break;
 	case ARM_OP_GROUP_PImmediate:
-		ARM_set_detail_op_imm(MI, OpNum, ARM_OP_PIMM,
+		ARM_set_detail_op_imm(MI, OpNum, ARM_OP_PIMM, 
```
trailing space

Rot127 commented on arch/ARM/ARMMapping.c:
```diff
@@ -520,13 +522,31 @@ static void add_cs_detail_general(MCInst *MI, arm_op_group op_group,
 			ARM_get_detail_op(MI, 0)->reg = Reg;
 			ARM_get_detail_op(MI, 0)->access = access;
+			cs_operand_encoding encoding = {
+				1, {reglist_encoding.indexes[piece_idx] + reg_idx}, {1}};
```
Please be more explicit here which member you set. Either via
```{ .bits = 1, .indices = {...} ...```
or by
```
cs_operand_encoding encoding = { 0 };
encoding..bits = 1;
...
```

orlitzky committed:
capstone.pc.in: re-add capstone include subdirectory via Cflags.

CMake installs the "common headers" for Capstone into a capstone/
subdirectory of the system's includedir:

  install(
    FILES ${HEADERS_COMMON}
    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/capstone
  )

The pkg-config file, on the other hand, omits the subdirectory:

    includedir=${prefix}/include
    Cflags: -I${includedir}

This means that consumers of the library would need to include the
"capstone/" prefix when including those headers. For example,

  #include <capstone/capstone.h>

rather than

  #include <capstone.h>

Neither is "wrong" -- it's a design choice, but one that breaks the
public API when it changes. This has happened before:

  https://github.com/capstone-engine/capstone/commit/0a39b785
  https://github.com/capstone-engine/capstone/commit/02e33f8b

And (for example) PHP was updated to reflect that, i.e. to work with
capstone-4.x:

  https://github.com/php/php-src/commit/ffc2a53a

But now in the 5.x series of Capstone, the path (or the -I flag,
relative to that path) has reverted,

  https://github.com/capstone-engine/capstone/commit/6656bcb6

breaking PHP once again. This commit re-fixes it by updating the
directory passed via Cflags.

Closes: https://github.com/capstone-engine/capstone/issues/1982

Rot127 commented on arch/ARM/ARMMapping.c:
```diff
@@ -1023,13 +1044,14 @@ static void add_cs_detail_template_1(MCInst *MI, arm_op_group op_group,
 		if (AlwaysPrintImm0)
 			map_add_implicit_write(MI, MCInst_getOpVal(MI, OpNum));
-
+		
```
Remove spaces

Rot127 commented on arch/ARM/ARMMapping.h:
```diff
@@ -63,7 +63,7 @@ static inline void add_cs_detail(MCInst *MI, int /* arm_op_group */ op_group,
 
 void ARM_set_detail_op_reg(MCInst *MI, unsigned OpNum, arm_reg Reg);
-void ARM_set_detail_op_sysreg(MCInst *MI, arm_sysreg SysReg, bool IsOutReg);
+void ARM_set_detail_op_sysreg(MCInst *MI, arm_sysreg SysReg, unsigned OpNum);
```
Please move the `OpNum` parameter after the `MI`. So all functions follow the same parameter pattern (`MI, OpNum, ...`).

Rot127 commented on include/capstone/arm.h:
```diff
@@ -411,13 +411,45 @@ typedef enum arm_op_type {
+	case ARM_OP_VPRED_R: return "ARM_OP_VPRED_R";
+	case ARM_OP_VPRED_N: return "ARM_OP_VPRED_N";
+	case ARM_OP_MEM: return "ARM_OP_MEM";
+	case ARM_OP_MEM | ARM_OP_IMM: return "ARM_OP_MEM | ARM_OP_IMM";
```
`case ARM_OP_MEM | ARM_OP_REG` is missing

Rot127 commented on include/capstone/capstone.h:
```diff
@@ -289,6 +289,36 @@ typedef struct cs_opt_skipdata {
+	/// The bit positions of each piece that form the full operand in order. If
+	/// there is only one piece then there is only one index as well. Likewise
+	/// if there are 4 pieces, there are 4 indexes and so on.
+	uint8_t indexes[8];
```
Move the `8` to a macro please. Same for the `64` below

Rot127 commented on include/capstone/capstone.h:
```diff
@@ -289,6 +289,36 @@ typedef struct cs_opt_skipdata {
 } cs_opt_skipdata;
 
+/// Provides information about an operand's encoding in the instruction
+typedef struct cs_operand_encoding {
```
General comment regarding the indices:

Please clarify that the bit indices _not_ necessarily match the indices in the ISA. Different architectures might label the left most bit as bit 0 others the right most bit.

Rot127 reviewed (commented):
Really nice work! Also thank you a lot to make `Update-Arch.sh` compatible with WSL! It is is so much better now.

Once the auto-sync ARM PR was reviewed and all requested fixes are pushed, please rebase this.

orlitzky committed:
capstone.pc.in: use https in the URL.

kabeor committed:
Merge pull request #2046 from orlitzky/pkgconfig-includes

Re-fix pkgconfig includedir/Cflags

kabeor committed:
update ChangeLog to 5.0.0-rc3

kabeor committed:
Merge branch 'next' of https://github.com/kabeor/capstone into next

kabeor committed:
Merge pull request #2047 from kabeor/next

update ChangeLog to 5.0.0-rc3

Rot127 commented on suite/auto-sync/Update-Arch.sh:
```diff
@@ -0,0 +1,279 @@
+  read -r
+  exit
+}
+print_and_pause() {
```
Please add newlines after each function.

Rot127 commented on suite/auto-sync/Update-Arch.sh:
```diff
@@ -0,0 +1,279 @@
+
+pause() {
+  echo "Press any key to exit"
+  read -r
```
In bash this doesn't wait for the user input. It needs a variable:
`read -r input`

AngelDev06 replied:
What should I use to wait for input, what I had before was nonstandard bash extentions.

Rot127 replied:
Doesn't matter. Only the argument must be set.

```suggestion
  read -r dummy_var
```

Rot127 commented on suite/auto-sync/Update-Arch.sh:
The script fails to run for me:

```
./suite/auto-sync/Update-Arch.sh ARM ~/repos/llvm-capstone/ 0000000
[x] /home/user/repos/llvm-capstone/ is an invalid path (it should have at least somewhere the "llvm-capstone/llvm" substr for it to work)
Press any key to exit

```

```
07:09 $ ./suite/auto-sync/Update-Arch.sh ARM ~/repos/llvm-capstone/llvm 0000000
[*] Create ./build/llvm_inc directory
[*] Create ./build/llvm_c_inc directory
[*] Create ./build/trans_out directory
[*] Create ./build/ts_libs directory
[*] Create ./build/diff_out directory
[*] Create symlink '../vendor/llvm_root -> /home/user/repos/llvm-capstone/llvm'.
ln: failed to create symbolic link '../vendor/llvm_root': No such file or directory
[x] llvm-tblgen not found at '/home/user/repos/llvm-capstone/llvm'
Press any key to exit

```

Besides, please update the help message. Should the llvm path point to `llvm-capstone` (to the project repo)? Or should it point to `llvm-capstone/llvm`?

AngelDev06 replied:
My bad, the error message is caused because originally I made it so that it fetches the llvm root by parsing the path. I changed the strategy as soon as I found out that git could do that for me (get the root of the repo), so now it's not really needed. Also I should have noted that now the second argument is a full path to the llvm-tblgen executable, instead of a path to the repo. I made that change because for me it was built in a folder that the script couldn't find. So now it accepts a full path to the executable and then by using git it gets the root of llvm-capstone as well. But as I have stated before all paths are now relative to llvm-capstone/llvm so after fetching the root it also appends the /llvm suffix. Anyway I pushed the changes.

Rot127 replied:
Ok. Please don't forget to update the helper message. Currently it still says it should be the path to the llvm-project repo dir.

Rot127 replied:
Also update the `suite/auto-sync/README.md`

AngelDev06 replied:
Done 👍

hamarituc committed:
Fix TriCore test

kabeor committed:
Merge pull request #2049 from hamarituc/tricore-test

Fix TriCore test

kabeor committed:
[workflows] update python-publish

kabeor committed:
Merge branch 'capstone-engine:next' into next

kabeor committed:
[workflows] update python-publish

kabeor committed:
Merge branch 'next' of https://github.com/kabeor/capstone into next

kabeor committed:
[bindings] fix twine issue when publish

kabeor committed:
[bindings] fix twine issue when publish

kabeor committed:
[workflows] update actions version

kabeor committed:
[workflows] remove manual publish

kabeor committed:
[bindings] Adapt python2 for setup.py

kabeor committed:
Merge pull request #2050 from kabeor/next

[workflows] Fix python-publish issue

kabeor committed:
Update version tag

kabeor committed:
Merge branch 'capstone-engine:next' into next

kabeor committed:
Merge pull request #2052 from kabeor/next

Update version tag

aquynh committed:
bindings: update binding consts

aquynh committed:
README: use https

Rot127 commented on suite/auto-sync/README.md:
````diff
@@ -0,0 +1,143 @@
+Run the updater
+
+```
+mkdir build
+cd build
````
```suggestion
```

Rot127 reviewed (commented):
LGTM after the change. Please note that it might take a while until this is merged (ARM, PPC and AArch64 should be done first. So it has some order). I would come back to you when we require to rebase it.

kabeor committed:
[workflows] Remove deprecated image versions

kabeor committed:
[workflows] Adjust matrix version

b1llow committed:
Fix capstone_test.c

b1llow committed:
Fix const_generator.py and tricore_const.py

kabeor committed:
Merge pull request #2057 from kabeor/next

[workflows] Remove deprecated image versions

arkup committed:
fix msvc static lib build error

arkup committed:
fix compile error capstone_dll unresolved external symbol _insn_find

b1llow committed:
Fix tricore.h

b1llow committed:
Update all *_const.py

adamjseitz committed:
Support repz prefix on X86 ret instructions

arkup committed:
fix static lib to compile with has_tricore

arkup committed:
fix msvc dll to compile with has_tricore, plus added CRT_SECURE_NO_WARNINGS

kabeor committed:
Merge pull request #2060 from imbillow/fix-capstone-test

Fix capstone_test.c #2059

kabeor committed:
Merge pull request #2063 from adamjseitz/repz-ret

kabeor committed:
Merge pull request #2061 from arkup/next

fix for msvc static lib, dll build error and warnings

ConnorRigby committed:
use calloc for cs_insn instead of malloc (#1972)

Signed-off-by: Connor Rigby <connorr@hey.com>

melynx committed:
Fix access bug for TEST op (#1511)

* Fix access bug for TEST op

Results for TEST instructions are discarded hence operands should have
no write access.

* Fix access bug for TEST (reduce)

* trigger new ci test

---------

Co-authored-by: kabeor <1597915586@qq.com>

kabeor committed:
Update ChangeLog to Version 5.0.0

XVilka committed:
Convert Tricore to Intel register syntax

XVilka committed:
Convert Tricore tests to Intel register syntax

peace-maker committed:
Normalize rc4 version string

/usr/local/lib/python2.7/dist-packages/setuptools/dist.py:476: UserWarning: Normalizing '5.0.0.rc4' to '5.0.0rc4'
  normalized_version,

peace-maker committed:
Search for correct versioned lib .5 suffix

Always use the major version instead of requiring manual updates to the
file names.

peace-maker committed:
Add CS_GRP_BRANCH_RELATIVE to python bindings

peace-maker committed:
Update Cython bindings

kabeor committed:
Merge pull request #2068 from peace-maker/cython_bindings

kabeor committed:
Merge pull request #2067 from XVilka/tricore-intel-syntax

Use Intel register syntax for Tricore

b1llow committed:
Add `#` prefix to tricore imm operand

b1llow committed:
Update tests and inc files

with llvm-capstone 78180b63f827ee38e23375e7a00825e848aa6956

kabeor committed:
Merge pull request #2070 from imbillow/tricore-fix

oleavr committed:
Revert "Mark cs_* as thread local to avoid race condition in multithreads" (#1933)

Having per-thread memory allocation routines is a major change in
behavior that likely broke a lot of applications.

This reverts commit bdab89496e654d043f556e27a68c2ac47dd571a7.

kabeor committed:
Update ChangeLog to Version 5.0.0

aquynh committed:
bindings: update Java consts for Arm64

aquynh committed:
ChangeLog

aquynh committed:
update pkgconfig.mk for 5.0

XVilka committed:
Remove outdated and irrelevant TODO

Rot127 committed:
Replace strncpy with memcpy to fix compiler warning.

kabeor committed:
Merge pull request #2075 from Rot127/sparc-strncpy

Replace strncpy with memcpy to fix compiler warning.

kabeor committed:
Merge pull request #2074 from XVilka/remove-todo

Remove outdated and irrelevant TODO

kabeor committed:
Update ChangeLog v5.0 Final

aquynh committed:
5.0-post1

aquynh committed:
python: update binding README

Rot127 commented on arch/ARM/ARMMapping.c:
```diff
@@ -887,6 +887,7 @@ static void add_cs_detail_general(MCInst *MI, arm_op_group op_group,
 			ARM_get_detail_op(MI, 2)->shift.type = ARM_SFT_LSL;
 			ARM_get_detail_op(MI, 2)->shift.value = ShAmt;
 		}
+		ARM_inc_op_count(MI);
```
This is done via `ARM_set_mem_access(MI, false);`

AngelDev06 replied:
I remember testing the THUMB2 instruction `ldrsht` and the memory operand was missing from the operands list. So I tried debugging it and it seemed like the operand counter wasn't incremented. I don't know if it was supposed to be incremented somewhere else but apparently it didn't

Rot127 replied:
Is also fixed by #1949. So you just have to resolve the conflict during the future rebase.

AngelDev06 committed:
Added memory operand format for ARM

XVilka committed:
[SuperH]: fix IMM format specifier in test (#2080)

Rot127 committed:
Architecture updater (auto-sync) - Updating ARM (#1949)

* Add auto-sync updater.

* Update Capstone core with auto-sync changes.

* Update ARM via auto-sync.

* Make changes to arch modules which are introduced by auto-sync.

* Update tests for ARM.

* Fix build warnings for make

* Remove meson.build

* Print shift amount in decimal

* Patch non LLVM register alias.

* Change type of immediate operand to unsiged (due to: #771)

* Replace all occurances of a register with its alias.

* Fix printing of signed imms

* Print rotate amount in decimal

* CHange imm type to int64_t to match LLVM imm type.

* Fix search for register names, by completing string first.

* Print ModImm operands always in decimal

* Use number format of previous capstone version.

* Correct implicit writes and update_flags according to SBit.

* Add missing test for RegImmShift

* Reverse incorrect comparision.

* Set shift information for move instructions.

* Set mem access for all memory operands

* Set subtracted flag if offset is negative.

* Add flag for post-index memory operands.

* Add detail op for BX_RET and MOVPCLR

* Use instruction post_index operand.

* Add VPOP and VPUSH as unique CS IDs.

* Add shifting info for MOVsr.

* Add TODOs.

* Add in LLVM hardcoded operands to detail.

* Move detail editing from InstPrinter to Mapping

* Formatting

* Add removed check.

* Add writeback register and constraints to RFEI instructions.

* Translate shift immediate

* Print negative immediates

* Remove duplicate invalid entry

* Add CS groups to instructions

* Fix write attriutes of stores.

* Add missing names of added instructions

* Fix LLVM bug

* Add more post_index flags

* http -> https

* Make generated functions static

* Remove tab prefix for alias instructions.

* Set ValidateMCOperand to NULL.

* Fix AddrMode3Operand operands

* Allow getting system and banked register name via API

* Add writeback to STC/LDC instructions.

* Fix (hopefully) last case where disp is negative and subtracted = true

* Remove accidentially introduced regressions

Rot127 committed:
Bumb manylinux version to 2014 (x86) and 2_28 (x86_64) repectivly.

manylinux1 is EOL since 2 years. Just like Python 3.6
Fixes wheel build.

Rot127 committed:
Run clang-format (#2093)

kabeor committed:
Merge pull request #2092 from Rot127/fix-build

[Workflows] Update manylinux docker images

XVilka committed:
ci: use proper job syntax for Windows

kabeor committed:
Merge pull request #2098 from XVilka/fix-windows-python-ci

ci: use proper job syntax for Windows

AngelDev06 commented:
Note: I rebased onto latest commit since I noticed that the ARM auto-sync pr was merged and I needed some of those latest changes for my own stuff (mostly for bug fixes). I am aware I may have to rebase again once more auto-sync archs get added and I will be willing to do that when the time comes

AngelDev06 committed:
Fixed bug with include order

AngelDev06 committed:
Fixed small bug in reglist encoding

Incorrect SP encoding

Rot127 commented on arch/ARM/ARMMapping.c:
```diff
@@ -401,7 +401,9 @@ static void ARM_add_reglist_reg_encoding(MCInst *MI, arm_reg reg)
-	unsigned RegBitNum = reg >= 73 ? reg - 73 : reg + 1;
+	unsigned RegBitNum = reg == ARM_REG_SP ? 13 :
+			     reg >= 73	       ? reg - 73 :
+						 reg + 1;
```
Please use no magic numbers, replace `73` with the enum name.
Also for `ARM_REG_SP` there should be some `ARM_REG_R13 = ARM_REG_SP` at the end of the enum. So please use this one.

AngelDev06 committed:
Replaced numbers with enum values

For better understanding purposes

XVilka commented:
Please fix fuzzing bugs as well:
```
fuzz_disasmnext: /src/capstonenext/arch/ARM/ARMMapping.c:1861: void ARM_set_detail_op_imm(MCInst *, unsigned int, arm_op_type, int64_t): Assertion `!(map_get_op_type(MI, OpNum) & CS_OP_MEM)' failed.
AddressSanitizer:DEADLYSIGNAL
=================================================================
==27==ERROR: AddressSanitizer: ABRT on unknown address 0x00000000001b (pc 0x7fb46332c00b bp 0x7fb4634a1588 sp 0x7ffc74ea3570 T0)
SCARINESS: 10 (signal)
    #0 0x7fb46332c00b in raise (/lib/x86_64-linux-gnu/libc.so.6+0x4300b) (BuildId: 1878e6b[47](https://github.com/capstone-engine/capstone/actions/runs/5631012546/job/15350252977?pr=2045#step:5:48)5720c7c51969e69ab2d276fae6d1dee)
    #1 0x7fb46330b858 in abort (/lib/x86_64-linux-gnu/libc.so.6+0x22858) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #2 0x7fb46330b728  (/lib/x86_64-linux-gnu/libc.so.6+0x22728) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #3 0x7fb46331cfd5 in __assert_fail (/lib/x86_64-linux-gnu/libc.so.6+0x33fd5) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #4 0x598546 in ARM_set_detail_op_imm /src/capstonenext/arch/ARM/ARMMapping.c:1861:2
    #5 0x596cbc in add_cs_detail_general /src/capstonenext/arch/ARM/ARMMapping.c
    #6 0x593145 in ARM_add_cs_detail /src/capstonenext/arch/ARM/ARMMapping.c:1808:2
    #7 0x7e011b in add_cs_detail /src/capstonenext/arch/ARM/ARMMapping.h:62:2
    #8 0x7da65b in printCImmediate /src/capstonenext/arch/ARM/ARMInstPrinter.c:1145:2
    #9 0x7d713b in printInstruction /src/capstonenext/arch/ARM/ARMGenAsmWriter.inc
    #10 0x7e1985 in printInst /src/capstonenext/arch/ARM/ARMInstPrinter.c
    #11 0x7e1985 in ARM_LLVM_printInstruction /src/capstonenext/arch/ARM/ARMInstPrinter.c:1895:2
    #12 0x586e58 in ARM_printer /src/capstonenext/arch/ARM/ARMMapping.c:437:2
    #13 0x56e8a6 in cs_disasm /src/capstonenext/cs.c:944:4
    #14 0x56c636 in LLVMFuzzerTestOneInput /src/capstonenext/suite/fuzz/fuzz_disasm.c:56:20
    #15 0x43ddc3 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
    #16 0x43d5aa in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:514:3
    #17 0x43f414 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:826:7
    #18 0x43f649 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:857:3
    #19 0x42ecaf in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:912:6
    #20 0x458302 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #21 0x7fb46330d082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #22 0x41f6ed in _start (build-out/fuzz_disasmnext+0x41f6ed)

DEDUP_TOKEN: raise--abort--
AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: ABRT (/lib/x86_64-linux-gnu/libc.so.6+0x4300b) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee) in raise
==27==ABORTING
MS: 0 ; base unit: 0000000000000000000000000000000000000000
0x3,0x33,0xed,0x1e,0x7f,
\0033\355\036\177
artifact_prefix='/tmp/tmpet2wpn1m/'; Test unit written to /tmp/tmpet2wpn1m/crash-7529d18d8f1d51daf81fcc7a43063df595571beb
Base64: AzPtHn8=
stat::number_of_executed_units: 4685
stat::average_exec_per_sec:     4685
stat::new_units_added:          0
stat::slowest_unit_time_sec:    0
stat::peak_rss_mb:              104
/github/workspace/build-out/fuzz_disasmnext -max_len=4096 -timeout=25 -rss_limit_mb=2560 -len_control=0 -seed=1337 -artifact_prefix=/tmp/tmpet2wpn1m/ -max_total_time=300 -print_final_stats=1 /github/workspace/cifuzz-corpus/fuzz_disasmnext >fuzz-1.log 2>&1
================== Job 1 exited with exit code 77 ============
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 1337
INFO: Loaded 1 modules   (38312 inline 8-bit counters): 38312 [0x1723b20, 0x172d0c8), 
INFO: Loaded 1 PC tables (38312 PCs): 38312 [0x14b1198,0x1546c18), 
INFO:    63359 files found in /github/workspace/cifuzz-corpus/fuzz_disasmnext
INFO: seed corpus: files: 63359 min: 1b max: 4096b total: 8870283b rss: 78Mb
fuzz_disasmnext: /src/capstonenext/arch/ARM/ARMMapping.c:1861: void ARM_set_detail_op_imm(MCInst *, unsigned int, arm_op_type, int64_t): Assertion `!(map_get_op_type(MI, OpNum) & CS_OP_MEM)' failed.
AddressSanitizer:DEADLYSIGNAL
=================================================================
==31==ERROR: AddressSanitizer: ABRT on unknown address 0x00000000001f (pc 0x7f623e53b00b bp 0x7f623e6b0588 sp 0x7ffd10872d50 T0)
SCARINESS: 10 (signal)
    #0 0x7f623e53b00b in raise (/lib/x86_64-linux-gnu/libc.so.6+0x4300b) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #1 0x7f623e51a858 in abort (/lib/x86_64-linux-gnu/libc.so.6+0x22858) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #2 0x7f623e51a728  (/lib/x86_64-linux-gnu/libc.so.6+0x22728) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #3 0x7f623e52bfd5 in __assert_fail (/lib/x86_64-linux-gnu/libc.so.6+0x33fd5) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #4 0x598546 in ARM_set_detail_op_imm /src/capstonenext/arch/ARM/ARMMapping.c:1861:2
    #5 0x596cbc in add_cs_detail_general /src/capstonenext/arch/ARM/ARMMapping.c
    #6 0x593145 in ARM_add_cs_detail /src/capstonenext/arch/ARM/ARMMapping.c:1808:2
    #7 0x7e011b in add_cs_detail /src/capstonenext/arch/ARM/ARMMapping.h:62:2
    #8 0x7da65b in printCImmediate /src/capstonenext/arch/ARM/ARMInstPrinter.c:1145:2
    #9 0x7d713b in printInstruction /src/capstonenext/arch/ARM/ARMGenAsmWriter.inc
    #10 0x7e1985 in printInst /src/capstonenext/arch/ARM/ARMInstPrinter.c
    #11 0x7e1985 in ARM_LLVM_printInstruction /src/capstonenext/arch/ARM/ARMInstPrinter.c:1895:2
    #12 0x586e58 in ARM_printer /src/capstonenext/arch/ARM/ARMMapping.c:437:2
    #13 0x56e8a6 in cs_disasm /src/capstonenext/cs.c:944:4
    #14 0x56c636 in LLVMFuzzerTestOneInput /src/capstonenext/suite/fuzz/fuzz_disasm.c:56:20
    #15 0x43ddc3 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
    #16 0x43d5aa in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:514:3
    #17 0x43f414 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:826:7
    #18 0x43f6[49](https://github.com/capstone-engine/capstone/actions/runs/5631012546/job/15350252977?pr=2045#step:5:50) in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:857:3
    #19 0x42ecaf in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:912:6
    #20 0x458302 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #21 0x7f623e[51](https://github.com/capstone-engine/capstone/actions/runs/5631012546/job/15350252977?pr=2045#step:5:52)c082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #22 0x41f6ed in _start (build-out/fuzz_disasmnext+0x41f6ed)

DEDUP_TOKEN: raise--abort--
AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: ABRT (/lib/x86_64-linux-gnu/libc.so.6+0x4300b) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee) in raise
==31==ABORTING
MS: 0 ; base unit: 0000000000000000000000000000000000000000
0x3,0x33,0xed,0x1e,0x7f,
\0033\355\036\177
artifact_prefix='/tmp/tmpet2wpn1m/'; Test unit written to /tmp/tmpet2wpn1m/crash-7[52](https://github.com/capstone-engine/capstone/actions/runs/5631012546/job/15350252977?pr=2045#step:5:53)9d18d8f1d51daf81fcc7a43063df59[55](https://github.com/capstone-engine/capstone/actions/runs/5631012546/job/15350252977?pr=2045#step:5:56)71beb
Base[64](https://github.com/capstone-engine/capstone/actions/runs/5631012546/job/15350252977?pr=2045#step:5:65): AzPtHn8=
stat::number_of_executed_units: 4[68](https://github.com/capstone-engine/capstone/actions/runs/5631012546/job/15350252977?pr=2045#step:5:69)5
stat::average_exec_per_sec:     4685
stat::new_units_added:          0
stat::slowest_unit_time_sec:    0
stat::peak_rss_mb:              103
2023-07-26 06:32:32,[80](https://github.com/capstone-engine/capstone/actions/runs/5631012546/job/15350252977?pr=2045#step:5:81)1 - root - INFO - Re
```

AngelDev06 commented:
Pretty sure this has to do with the commit onto which I rebased. Shouldn't be an issue on future rebase

Rot127 commented:
I'll do a second round of review soon.
But do you think we can add a reference to each `cs_arm_op` which piece(s) of encoding are its source?

It would be nice, because when printing the details we could also print the variable names of each operand.

If every encoding piece has an `OpNum` assigned to it, the code could simply be moved to `ARM_set_detail_op_<reg/imm/mem>()`, right?

AngelDev06 commented:
Each OpNum doesn't necessarily correspond to a single piece. 
That depends on how that specific operand is encoded. 
Some operands don't have all of their bits next to each other and instead have some of them at one position and the rest on another. My favourite example is the [encoding T1 of mov register](https://developer.arm.com/documentation/ddi0406/c/Application-Level-Architecture/Instruction-Details/Alphabetical-list-of-instructions/MOV--register--Thumb-?lang=en) which has one bit of the destination register operand (Rd) at position 7 while the other 3 are at the begining (positions 0-2). 
So in that case OpNum would correspond to **two pieces of the same operand**. And yes because each OpNum corresponds to pieces of the same operand I have moved all of the mapping operations (regarding the encoding) to `ARM_set_detail_op_<reg/imm/mem>()` except for a few exceptions such as the reglist operand. 
Because capstone adds an operand entry for each individual register in the reglist I had to make the encoding specific to each of these registers and not the entire reglist. Also yes in each `cs_arm_op` entry all of the operand pieces provided are part of that specific operand so you know to which operand those pieces belong to.

There is only one problem and that's why I added my second commit. 
Since memory operands are complicated and therefore the `*.td` files don't provide extra info about the encoding of their suboperands, I can only get the encoding of the **entire** memory operand. That's ok considering that capstone does add an entry for the entire memory operand and sets its type as `ARM_OP_MEM`, however some user might want to know whether a piece provided corresponds to the base register of the memory operand or to the index one, and so on. To add this information, I also added a field to `arm_op_mem` named format which holds some hardcoded enum value specifying how the memory operand is formatted.
This tells you exactly which piece should be the base register, which one the displacement or index register etc.

Rot127 commented on include/capstone/capstone.h:
```diff
@@ -290,6 +290,40 @@ typedef struct cs_opt_skipdata {
+	/// The bit positions of each piece that form the full operand in order. If
+	/// there is only one piece then there is only one index as well. Likewise
+	/// if there are 4 pieces, there are 4 indexes and so on.
+	/// Also note that these indices DON'T necessarily match the indices in the ISA
```
I think we should always save bits at the same index as the ISA does. Otherwise there is no reference point for people.

Rot127 commented:
@AngelDev06 Just rebased https://github.com/capstone-engine/capstone/pull/2026. You shouldn't get conflicts anymore.

Rot127 commented:
Ok, so there are some more changes coming (nothing ARM related). You might as well also just wait with the rebase. I can give you the go then some when later this month, if this is fine for you.

Rot127 commented:
Is it fine for you, if I cherry-pick your commits into the auto-sync PR? It is easier to work on it this way. Also it would probably be merged faster.
You can still work on it by opening PRs on the auto-sync [branch](https://github.com/Rot127/capstone/tree/auto-sync-aarch64)

AngelDev06 commented:
I might be on holidays for about two weeks starting from 15th of this month so I might not be able to rebase. But in the worst case senario, I will be available either at the end of this month or at the start of the next one. Yes you can cherry pick, although I can still rebase onto your changes of the arm64 pr if you want. 

Also note that I will try to fix the indexes to be according to the ARM documentation & also open a brand new pr on llvm-capstone that includes my changes.

Up to you if you want me to rebase now or wait for more changes to come as you mentioned.

Rot127 commented:
My problem is that `llvm-capstone` is out of sync and generates newer tables (with the encoding). And I am tiered of cherry picking the right commits into other branches. Also I'd like to have your encoding work merged end of the month (if @kabeor finds time to review it).
If you want to change little things they can be done after the merge as well.

So I would cherry pick them then. Especially if you are on vacation (wish you a nice one btw).

> also open a brand new pr on llvm-capstone that includes my changes.

If you manage to do this before your vacation it would be great. Or does it change a lot?

> Also note that I will try to fix the indexes to be according to the ARM documentation

Sure. If this happens after https://github.com/capstone-engine/capstone/pull/2026 it should be fine as well.
Better check before if the indices in LLVM maybe already match the ones in the ISA. Then we just fix the doc string.

AngelDev06 commented:
> My problem is that `llvm-capstone` is out of sync and generates newer tables (with the encoding). And I am tiered of cherry picking the right commits into other branches. 

I can do some small patches to my current clone of `llvm-capstone` to make them sync to what capstone needs right now and push the newer commits so hopefuly you won't have to do any manual work. After all the only thing I generate is `ARMGenCSMappingInsn.inc` & `ARMGenCSMappingInsnOp.inc`

> Also I'd like to have your encoding work merged end of the month (if @kabeor finds time to review it). If you want to change little things they can be done after the merge as well.

Sure. Is my work going to be included as part of #2026 or merged directly from here?

> If you manage to do this before your vacation it would be great. Or does it change a lot?

~~It doesn't change a lot. I can get it done today.~~ Done!

> Sure. If this happens after #2026 it should be fine as well. Better check before if the indices in LLVM maybe already match the ones in the ISA. Then we just fix the doc string.

~~This can be done now if needed. It's a very small patch which I should have done before my pr on `llvm-capstone`.~~ Also done

AngelDev06 committed:
Generated new tables to make operand encoding indexes based on ISA

AngelDev06 committed:
Made `ARMMapping.c` comply with latest changes

Regarding the encoding index changes

AngelDev06 committed:
Removed outdated part of encoding docstring

AngelDev06 committed:
Removed tmp from `insn_map`

AngelDev06 committed:
Made the encoding of base & offset for post indexed intructions to be merged

Rot127 commented:
Sorry for letting this PR stall for so long. I looked into it quite a while, and we discussed this internally as well.
Unfortunately, we won't merge it.

The problem is not, that we don't need it, but that the current implementation doesn't work for other architectures.
This is not really a problem with your design, but first and foremost due to the flawed definitions in LLVM.

As you know, you had to do quite some manipulations of the encodings in `ARMMapping.c`, so the result is useable.
Additionally, I couldn't manage to have a one to one mapping from the ISA encoding classes to the one we have in LLVM/Capstone
(see: https://github.com/capstone-engine/capstone/commit/cff87a38061edb2a8968ef37cef6e0642bef6332. Never succeeded with it).

While these points would be acceptable for a single architecture module, we cannot do it for all the others.
If we add the instruction encodings as a feature, it must fit the paradigm of being easy to update and be comparable to LLVM output and the ISA.
But I couldn't find a way to make this possible for all other modules.

Due to the faulty encoding information in LLVM (and ARM is actually not the worst here), this is not maintainable without extra effort.
We lack maintainers as you probably know, so adding a relatively niche feature, which would consume more time is currently not an option, unfortunately.

However, it would be great if you would keep this feature up to date in your repository, so we can refer to it if other people need it for ARM.
I will move the relevant commits in llvm-capstone to another branch as well.

Also, we plan to add a generator for the SAIL definitions of several architectures. They should have way less flawed encoding information.
So once this is done, I would be delighted if you want to add it again.

Thank you a lot for the effort you put into it! And I am sorry I figured out too late, that it is not transferable to the other architecture modules.

AngelDev06 commented:
It's understandable.
I can tell that it was quite painful to try and get all the encodings correct when there were multiple inaccuracies and missing definitions in the `.td` files making automation a lot harder than it should have been. 
Considering that all of that was just for ARM I can imagine how time-consuming it can be to get it done for each architecture. 

Sadly I can't easily keep up my fork to date since I barely have any time for my own project, so I will probably just leave it as is for the time being.

I truly hope this project gets more maintainers in the future since it has a lot of potential and I personally, couldn't find a better disassembler for my needs.
Despite all that thanks for your effort in trying to integrate my changes into the library.
